django/contrib/admin/static/admin/js/theme.js

```diff
@@ -1,56 +1,51 @@
+            mode = "auto";
         }
+        document.documentElement.dataset.theme = mode;
+        localStorage.setItem("theme", mode);
+    }
 
-        function cycleTheme() {
-            const currentTheme = localStorage.getItem("theme") || "auto";
-            const prefersDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
+    function cycleTheme() {
+        const currentTheme = localStorage.getItem("theme") || "auto";
+        const prefersDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
 
-            if (prefersDark) {
-                // Auto (dark) -> Light -> Dark
-                if (currentTheme === "auto") {
-                    setTheme("light");
-                } else if (currentTheme === "light") {
-                    setTheme("dark");
-                } else {
-                    setTheme("auto");
-                }
+        if (prefersDark) {
+            // Auto (dark) -> Light -> Dark
+            if (currentTheme === "auto") {
+                setTheme("light");
+            } else if (currentTheme === "light") {
+                setTheme("dark");
             } else {
-                // Auto (light) -> Dark -> Light
-                if (currentTheme === "auto") {
-                    setTheme("dark");
-                } else if (currentTheme === "dark") {
-                    setTheme("light");
-                } else {
-                    setTheme("auto");
-                }
+                setTheme("auto");
+            }
+        } else {
+            // Auto (light) -> Dark -> Light
+            if (currentTheme === "auto") {
+                setTheme("dark");
+            } else if (currentTheme === "dark") {
+                setTheme("light");
+            } else {
+                setTheme("auto");
             }
         }
+    }
 
-        function initTheme() {
-            // set theme defined in localStorage if there is one, or fallback to auto mode
-            const currentTheme = localStorage.getItem("theme");
-            currentTheme ? setTheme(currentTheme) : setTheme("auto");
-        }
-
-        function setupTheme() {
-            // Attach event handlers for toggling themes
-            const buttons = document.getElementsByClassName("theme-toggle");
-            Array.from(buttons).forEach((btn) => {
-                btn.addEventListener("click", cycleTheme);
-            });
-            initTheme();
-        }
+    function initTheme() {
+        // set theme defined in localStorage if there is one, or fallback to auto mode
+        const currentTheme = localStorage.getItem("theme");
+        currentTheme ? setTheme(currentTheme) : setTheme("auto");
+    }
 
-        setupTheme();
+    window.addEventListener('load', function(_) {
+        const buttons = document.getElementsByClassName("theme-toggle");
+        Array.from(buttons).forEach((btn) => {
+            btn.addEventListener("click", cycleTheme);
+        });
     });
+
```

Comment:
```suggestion
    // reset theme if auto mode activated and os preferences have changed
    window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", function (e) {
        prefersDark = e.matches;
        initTheme();
    })
```
Small addition to make sure it's switching correctly in auto mode.

Reply:
🤩 
> To perform a one-time, instantaneous check to see if the document matches the media query, look at the value of the [matches](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/matches) property, which will be true if the document meets the media query's requirements.
> https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia

Reply:
I think this might already been handled by the CSS

https://github.com/django/django/assets/46227124/cae006d9-abff-45f8-be0f-c8a2cb88346c

Reply:
It does but the initTheme is not triggered initially so you have the initial issue at the first load of the admin if you are in auto mode.

Reply:
@betaflag Do you have time to keep working on this?

Reply:
@felixxm I just got back from the holidays, I should be able to complete this today!

Reply:
Great, thanks :gift:

Reply:
@sabderemaneI'm not able to reproduce any flickering with auto-mode and toggling OS theme preference. Can you share the steps? 

> initTheme is not triggered initially

[It is actually](https://github.com/django/django/pull/17625/files#diff-df08791ae35200fa9e04430a08b89b37d9f6164d53b4effeaaae0e0bc73fd49bR50)

Reply:
I can't reproduce it anymore, I don't remember the exact steps I have done, maybe it was just caching issue, so closing for now.